### PR TITLE
[SYSSETUP][BOOTDATA] Make Server default again and write Service Pack

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

- Make "ReactOS Server" the default product option again instead of "ReactOS Workstation".
- Write "Service Pack" info onto registry.
- Add `ProductOption` option to `bootcd` `unattend.inf`.
- Delete `IDC_PRODUCT_SUITE` and `IDC_PRODUCT_TYPE` controls.

JIRA issue: [CORE-17028](https://jira.reactos.org/browse/CORE-17028)

after

### CODE REVIEW

File: dll/win32/syssetup/wizard.c [hunk 1]
```diff
@@ -499,7 +521,7 @@ ProductPageDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
             LoadStringW(hDllInstance, IDS_PRODUCTWORKSTATIONNAME, szText, _countof(szText));
             SendDlgItemMessageW(hwndDlg, IDC_PRODUCT_OPTIONS, CB_ADDSTRING, 0, (LPARAM)szText);
 
-            SendDlgItemMessageW(hwndDlg, IDC_PRODUCT_OPTIONS, CB_SETCURSEL, PRODUCT_OPTION_WORKSTATION, 0);
+            SendDlgItemMessageW(hwndDlg, IDC_PRODUCT_OPTIONS, CB_SETCURSEL, PRODUCT_OPTION_SERVER, 0);
             OnChooseWorkstation(hwndDlg);
```
Comment (KyleKatarn31) [inline dll/win32/syssetup/wizard.c:525]: Is OnChooseWorkstation still relevant here if Server is default ?
I'd suggest OnChooseServer(hwndDlg); instead

File: dll/win32/syssetup/wizard.c [hunk 2]
```diff
@@ -448,6 +452,24 @@ DoWriteProductOption(PRODUCT_OPTION nOption)
 
     RegCloseKey(hKey);
 
+    error = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szWindowsNT, 0, KEY_WRITE, &hKey);
+    if (error)
+        return FALSE;
+
+    /* write CSDVersion value */
+    if (nOption == PRODUCT_OPTION_WORKSTATION)
```
Comment (Extravert-ir) [inline dll/win32/syssetup/wizard.c:460]: The function has two places where it checks `nOption` - one switch above and this if condition. Why so complicated?

**Comment by Extravert-ir** [comment]
By the way, this option should be added to unattend.inf file for automation

File: dll/win32/syssetup/wizard.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

File: dll/win32/syssetup/wizard.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: Simplified code.

File: boot/bootdata/livecd/unattend.inf [hunk 3]
```diff
@@ -11,3 +11,7 @@ UnattendSetupEnabled = no
 ; see hivesys.inf for available languages
 LocaleID = 409
 
+; set product option
+; 0: ReactOS Server
+; 1: ReactOS Workstation
+ProductOption = 0
```
Comment (learn-more) [inline boot/bootdata/livecd/unattend.inf:17]: At the moment livecd does not have handling of product option at all, so it is not needed in the unattend.inf for the livecd either :)

Comment (katahiromz) [reply boot/bootdata/livecd/unattend.inf]: Reverted.

**Comment by katahiromz** [comment]
broken
Why is `CSDVersion` broken?

File: dll/win32/syssetup/wizard.c [hunk 4]
```diff
@@ -410,69 +410,140 @@ DoWriteProductOption(PRODUCT_OPTION nOption)
...
+    RegCloseKey(hKey1);
+    RegCloseKey(hKey2);
+
+    return !error1 && !error2;
+}
+
+static BOOL
+DoWriteProductOption(PRODUCT_OPTION nOption)
+{
+    return WriteProductSuiteAndType(nOption) &&
```
Comment (Extravert-ir) [inline dll/win32/syssetup/wizard.c:519]: Why the separation? I don't get it - isn't it simpler to just write all registry values in a single `case` or `if` statement?

Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: United.

**Comment by katahiromz** [comment]
Now Service Pack is done. The remainders are OS version and build number.

File: dll/win32/syssetup/wizard.c [hunk 5]
```diff
@@ -390,89 +391,217 @@ AckPageDlgProc(HWND hwndDlg,
...
+static const WCHAR s_szProductOptions[] = L"SYSTEM\\CurrentControlSet\\Control\\ProductOptions";
+static const WCHAR s_szRosVersion[] = L"SYSTEM\\CurrentControlSet\\Control\\ReactOS\\Settings\\Version";
+static const WCHAR s_szWindowsNT[] = L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";
+static const WCHAR s_szControlWindows[] = L"SYSTEM\\CurrentControlSet\\Control\\Windows";
+static const WCHAR s_szCurrentBuildNumber[] = L"CurrentBuildNumber";
+static const WCHAR s_szCurrentVersion[] = L"CurrentVersion";
+
 static BOOL
-DoWriteProductOption(PRODUCT_OPTION nOption)
+WriteProductOptionServer(void)
```
Comment (HBelusca) [inline dll/win32/syssetup/wizard.c:402]: You could simplify the code and combine this function with the other one for Workstation, if you also had an array like:
```
{
    {L"ServerNT", L"Terminal Server", L"5.2", L"3790", 0x200, TRUE},
    {L"WinNT", L"", L"5.1", L"2000", 0x300, TRUE} 
}
```
of versions with data.

File: dll/win32/syssetup/wizard.c [hunk 6]
```diff
@@ -390,89 +391,217 @@ AckPageDlgProc(HWND hwndDlg,
...
+
+        /* write WindowsNT CurrentVersion */
+        cbData = sizeof(L"5.1");
+        error = RegSetValueExW(hKey, s_szCurrentVersion, 0, REG_SZ, (const BYTE *)L"5.1", cbData);
+        if (error)
+            break;
+
+        /* write WindowsNT CurrentBuildNumber */
+        cbData = sizeof(L"2000");
+        error = RegSetValueExW(hKey, s_szCurrentBuildNumber, 0, REG_SZ, (const BYTE *)L"2000", cbData);
```
Comment (SergeGautherie) [inline dll/win32/syssetup/wizard.c:543]: XPsp3 reports `2600`.
Is `2000` the actual value on S03sp2?

Comment (HBelusca) [reply dll/win32/syssetup/wizard.c:543]: You are correct concerning 2600:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/appcompat/shims/layer/versionlie.c;hb=98c17d3120b66c6c9507474cb5c3db1ca7b36ed3#l37

Comment (Doug-Lyons) [reply dll/win32/syssetup/wizard.c:543]: W2K3_SP2_Version

See Above, It should be 3790.

Comment (HBelusca) [reply dll/win32/syssetup/wizard.c:543]: @Doug-Lyons : the comment was about windows XP, not about windows 2003.

**Comment by HBelusca** [comment]
@katahiromz : Please check whether these version values remain present after installation and reboot of ReactOS.

File: dll/win32/syssetup/wizard.c (hunk 6 above)
Comment (Doug-Lyons) [reply dll/win32/syssetup/wizard.c:543]: Sorry, I saw the question from @SergeGautherie about S03sp2 and I was trying to clarify it.

Comment (SergeGautherie) [reply dll/win32/syssetup/wizard.c:543]: Doug, this line is about S03sp2 reporting as workstation.
Is that what you tested?

**Comment by katahiromz** [comment]
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmsysini.c;hb=4246b64dcb8ae6478f516bfe934d5c749b77deb0#l2116
`CmpSetVersionData` is writing registry of OS version info.

https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ex/init.c;hb=4246b64dcb8ae6478f516bfe934d5c749b77deb0#l42
`NtMajorVersion`, `NtMinorVersion` and `NtBuildNumber` seem fixed values. Do we have to modify the kernel?

File: dll/win32/syssetup/wizard.c (hunk 5 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

File: dll/win32/syssetup/wizard.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: Changed to `2600`.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_11_05_2020_09_08_58
`CurrentVersion` and `CurrentBuildNumber` won't change.

**Comment by HBelusca** [comment]
> Do we have to modify the kernel?

No we MUST not. All this version play-around should be only for user-mode applications.
That's why at some point I was thinking with @learn-more that we may also use (user-mode) shims to intercept calls to registry APIs and modify in flight the returned data in case the app(s) in question are looking at the reported version(s) by calling these APIs.

**Comment by katahiromz** [comment]
I don't know the mechanism of shim... How can I fix `CurrentVersion` and `CurrentBuildNumber`?

File: dll/win32/syssetup/wizard.c [hunk 7]
```diff
@@ -390,89 +391,152 @@ AckPageDlgProc(HWND hwndDlg,
...
+    DWORD dwValue, cbData;
+    const PRODUCT_OPTION_DATA *pData = &s_ProductOptionData[nOption];
+    ASSERT(0 <= nOption && nOption < _countof(s_ProductOptionData));
 
-    error = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szProductOptions, 0, KEY_WRITE, &hKey);
-    if (error)
-        return FALSE;
-
-    switch (nOption)
+    do
```
Comment (Extravert-ir) [inline dll/win32/syssetup/wizard.c:441]: Why not just use a common pattern `goto Error;`? :)

Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

**Comment by katahiromz** [comment]
@HBelusca You can modify this PR.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_11_05_2020_09_08_58
`CurrentVersion` and `CurrentBuildNumber` won't change.

**Comment by katahiromz** [comment]
Ignoring `CurrentVersion` and `CurrentBuildNumber`...

**Comment by katahiromz** [comment]
@JoachimHenze Retest this PR please!

**Comment by katahiromz** [comment]
@JoachimHenze Can you retest this PR? Thanks.

File: dll/win32/syssetup/wizard.c [hunk 8]
```diff
@@ -437,42 +438,101 @@ DoWriteProductOption(PRODUCT_OPTION nOption)
...
+OnChooseOption(HWND hwndDlg, PRODUCT_OPTION option)
 {
     WCHAR szText[256];
 
-    SetDlgItemTextW(hwndDlg, IDC_PRODUCT_SUITE, L"");
-    SetDlgItemTextW(hwndDlg, IDC_PRODUCT_TYPE, L"WinNT");
+    switch (option)
+    {
+        case PRODUCT_OPTION_SERVER:
+            SetDlgItemTextW(hwndDlg, IDC_PRODUCT_SUITE, L"Terminal Server");
```
[the remark this answers is not in the document]
Comment (KyleKatarn31) [reply dll/win32/syssetup/wizard.c]: Sorry these comments were prepared but not sent.... Maybe obsolete now (sorry)

**Comment by KyleKatarn31** [comment]
Sorry these comments were prepared but not sent.... Maybe obsolete now (sorry)

Do you need / want tests from me ?

**Comment by katahiromz** [comment]
@KyleKatarn31 Please test, I beg you.

**Comment by KyleKatarn31** [comment]
> 
> 
> @KyleKatarn31 Please test, I beg you.

I don't have VS2010 Ultimate to fully replay the ticket JIRA. I'll test anyway for standard operation and i'll post my findings here, if any.

**Comment by KyleKatarn31** [comment]
My comments after testing :
- Is it intentionnal that "Product Suite" is empty when Workstation is selected during installation ?
Installed as a Server
- In "System Properties" > System settings > Show as workstation is set by default. Is it intentionnal ?
Installed as a Workstation
- In "System Properties" > System settings > Claims that ReactOS is "presented as a Server OS". Is it still relevant then ?

**Comment by Extravert-ir** [comment]
VS 2010 installs now with ProductOption=Workstation.
What I don't like by the way, it spamming with
```
(dll\ntdll\rtl\version.c:62) We modified the reported OS from NtProductServer to NtProductWinNt
```
messages. Why this warning was put there at all?

File: dll/win32/syssetup/wizard.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply dll/win32/syssetup/wizard.c]: Is it localized on Windows? (Give an example.)

[the remark this answers is not in the document]
Comment (KyleKatarn31) [reply dll/win32/syssetup/wizard.c]: Does it has to be localized only if Windows does ? Seriously ?
Please take time to look at the rest of this module, is it worth keeping only a couple of strings in plain english while the rest is localized ?

[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/syssetup/wizard.c]: Personally I would remove this "ProductSuite" and "ProductType" thing from the installer UI, but keep only the description. Indeed what it shows in these two fields are the parameter strings stored in the registry values for Windows compatibility, and it doesn't make much sense from the point of view of the installer to show them.

**Comment by HBelusca** [comment]
> My comments after testing :
> 
>     * Is it intentionnal that "Product Suite" is empty when Workstation is selected during installation ?
> 
>     * Installed as a Server >> In "System Properties" > System settings > Show as workstation is set by default. Is it intentionnal ?
> 
>     * Installed as a Workstation >> In "System Properties" > System settings > Claims that ReactOS is "presented as a Server OS". Is it still relevant then ?

See my comment above (i.e. useless to show ProductType and ProductSuite registry value strings in the installer UI).

File: dll/win32/syssetup/wizard.c [hunk 9]
```diff
@@ -437,42 +438,101 @@ DoWriteProductOption(PRODUCT_OPTION nOption)
...
-    WCHAR szText[256];
+    static const WCHAR s_szWindowsNT[] = L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";
+    static const WCHAR s_szServicePack2[] = L"Service Pack 2";
+    static const WCHAR s_szServicePack3[] = L"Service Pack 3";
+    HKEY hKey;
+    LONG error;
+    DWORD cbData;
+
+    error = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szWindowsNT, 0, KEY_WRITE, &hKey);
+    if (error)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c:474]: `DPRINT1`'s inserted.

File: dll/win32/syssetup/wizard.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: Deleted.

File: dll/win32/syssetup/wizard.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (KyleKatarn31) [reply dll/win32/syssetup/wizard.c:474]: OK, thanks. This one to be resolved.

File: dll/win32/syssetup/wizard.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (KyleKatarn31) [reply dll/win32/syssetup/wizard.c]: I agree with @HBelusca on this from a UI/UX perspective.
